### ISSUE TITLE
[Accordion] Incorrect use of aria-hidden attribute when data-toggle is expanded

### ISSUE THREAD

**Issue by tstermitz** [zurb/foundation-sites#11629]
Similar to  #9913 : [tabs] Incorrect use of aria-hidden attribute

## What should happen?
**Level Access suggests:** 
The aria-hidden property should be treated as a binary property: either applied on an element with a value of "true", or not applied at all. When the structures are meant to be exposed for User interaction, the aria-hidden property should be removed entirely, and then reapplied with a value of "true" when the section is collapsed.

## What happens instead?
When data-toggle is expanded, the panel opens, and the property aria-hidden="true" is changed to aria-hidden="false" 

## Possible Solution
Remove the entire property aria-hidden="false" when open, and recreate it when closed.

## Test Case and/or Steps to Reproduce (for bugs)

How to reproduce:
1. Use the data-toggle on a button to open a div
2. Note that aria-hidden property on the div now says aria-hidden="false"

## Context
[From the W3C website:](https://www.w3.org/TR/wai-aria-1.1/#aria-hidden)

Note: Authors are advised to avoid using aria-hidden="false" with styles or attributes that have historically prevented rendering in all modalities, such as display:none or visibility:hidden in CSS, or the hidden attribute in HTML 5. At the time of this writing, aria-hidden="false" is known to work inconsistently when used in conjunction with such features.

## Your Environment
- FOUNDATION_VERSION = '6.3.1'
- Drupal 8.6.2

## Checklist
- [ x] I have read and follow the CONTRIBUTING.md document.
- [x ] There are no other issues similar to this one.
- [x ] The issue title and template are correctly filled.

**Comment by DanielRuf** [comment]
Not sure what issues W3C means but people should not use the attribute selector in this case as we have special classes.

Imo this is not an incorrect usage but browsers treat it differently and it's up to the developer if he / she uses it.

**Comment by DanielRuf** [comment]
I guess it is a misundedsfanding.

See the `values` table at the bottom:

Values:
Value	Description
false	The element is exposed to the accessibility API as if it was rendered.
true	The element is hidden from the accessibility API.
undefined (default)	The element's hidden state is determined by the user agent based on whether it is rendered.

**Comment by tstermitz** [comment]
Confusion on this issue is warranted, as you have to make a close reading of  the W3C docs; sort of like sussing out a double-negative. I wouldn't be reporting this if Level Access hadn't flagged it.

If I understand it, the problem is that you shouldn't combine aria-hidden="false" with the use of display: none or visibility: hidden or the hidden attribute. It's not just redundant, but it "is known to work inconsistently". You can and should use aria-hidden="true" when the element is hidden, but should not show aria-hidden at all when the element is visible. 

If you think about it, why would you need to indicate that something is not hidden? 

Issue #9913 regarding tabs is the clearest explanation I've seen of the problem, and a fix was created for aria-hidden="false" in the tabs module.

**Comment by DanielRuf** [comment]
> but should not show aria-hidden at all when the element is visible

This is not what I read there, see the values table.

In css, we should target "aria-hidden=true" and hide them by default or if the attr is missing or set to `false`.

**Comment by DanielRuf** [comment]
The `"false"` is not wrong afaik.

**Comment by DanielRuf** [comment]
See https://www.w3.org/WAI/ARIA/track/actions/1322

**Comment by DanielRuf** [comment]
aria-hidden="true" only hides it for screen readers. So we have to react in the normal mode and hide it too (if we want to dk thhs) - which is correct.

Also the value is "false", "true" or another string or empty string.

Please show us the line which is problematic in our CSS in your opinion.

**Comment by DanielRuf** [comment]
> shouldn't combine aria-hidden="false" with the use of display: none or visibility: hidden or the hidden attribute.

Which makes not much sense, hiding something in CSS and unhiding something for screen readers. I think you mean "true".

**Comment by DanielRuf** [comment]
> When data-toggle is expanded, the panel opens, and the property aria-hidden="true" is changed to aria-hidden="false"

Which is correct.
Please check the devtools un Chrome, the accessibility tool of it and check the value of the aria-hidden when it is set to "false", "" and without it.

**Comment by DanielRuf** [comment]
> When data-toggle is expanded, the panel opens, and the property aria-hidden="true" is changed to aria-hidden="false"

Which is correct.
Please check the devtools in Chrome, the accessibility tool of it and check the value of the aria-hidden attribute there when it is set to "false", "" and without it.

**Comment by tstermitz** [comment]
> Which is correct.

Not according to Level Access... "It's know to work inconsistently" (I guess for some browsers or whatever).

Show this accordion drop-down only for screen readers? Maybe there is a use case...

The specs don't say that aria-hidden="false" is "wrong" in general, but in this case, it is called out as "incorrect usage". There is no reason to tell an aria device "Yes, I'm really not hidden", you just need the absence of aria-hidden="true".

The problem is that foundation sees the data-toggle parameter and automatically adds the aria properties, in the toggle button and the chosen Div. 

**Maybe I'm looking at the wrong thing.** Are you saying that we should not use data-toggle for making accordion drop-downs visible and not visible? We should use some other foundation trigger, like tab?

**Comment by DanielRuf** [comment]
> The specs don't say that aria-hidden="false" is "wrong" in general, but in this case, it is called out as "incorrect usage". There is no reason to tell an aria device "Yes, I'm really not hidden", you just need the absence of aria-hidden="true".

Please reread the previozs comments and the values table on the W3C website.

**Comment by DanielRuf** [comment]
> The specs don't say that aria-hidden="false" is "wrong" in general, but in this case, it is called out as "incorrect usage". There is no reason to tell an aria device "Yes, I'm really not hidden", you just need the absence of aria-hidden="true".

Please reread the previous comments and the values table on the W3C website.

**Comment by DanielRuf** [comment]
> Maybe I'm looking at the wrong thing. Are you saying that we should not use data-toggle for making accordion drop-downs visible and not visible? We should use some other foundation trigger, like tab?

Depends on your needed UI.

> The problem is that foundation sees the data-toggle parameter and automatically adds the aria properties, in the toggle button and the chosen Div.

I see no problem there. "false" is like the default state and correct.

**Comment by DanielRuf** [comment]
> Maybe I'm looking at the wrong thing. Are you saying that we should not use data-toggle for making accordion drop-downs visible and not visible? We should use some other foundation trigger, like tab?

Depends on your needed UI.

> The problem is that foundation sees the data-toggle parameter and automatically adds the aria properties, in the toggle button and the chosen Div.

I see no problem there.

We've got a11y audits on big projects and so far no user said that this is not valid and would not allow them to pass the audit.

Also I think the "fix" that you've linked was rushed and not checked / peer-reviewed before.

**Comment by DanielRuf** [comment]
Also see https://github.com/zurb/foundation-sites/pull/9916#issuecomment-295309715 and https://github.com/zurb/foundation-sites/pull/9972

Marius / Owlbertz is our a11y professional.

**Comment by tstermitz** [comment]
Thank you for all your attention. 

I would be most happy if your perspective is the correct one.

**Comment by DanielRuf** [comment]
cc @Owlbertz

**Comment by Owlbertz** [comment]
Hi,

according to [the W3C spec](https://www.w3.org/TR/wai-aria-1.0/states_and_properties#aria-hidden):
> If an element is only visible after some user action, authors MUST set the aria-hidden attribute to true. When the element is presented, authors MUST set the aria-hidden attribute to false or remove the attribute, indicating that the element is visible

Therefore - as already mentioned above - setting it to `false` seems perfectly fine according to the spec. I do agree that there is of course not the need to add `aria-hidden="false"` as stated in https://github.com/zurb/foundation-sites/issues/11629#issuecomment-447166653, but for the sake of simplification of the code I would say removing the attribute instead of just applying the variable value (i.e. `false`) would cause a more confusing/complex code base, which is not necessary.
Therefore I would say that there is no need in changing the current implementation and keep the `aria-hidden="false"`.

**Comment by tstermitz** [comment]
Thank you Marius for your attention. 

I read the W3C specs the same way you do, and I would not be raising the issue except that it was identified by an audit as a "medium-level" fault by [Level Access](https://www.levelaccess.com)

I can accept this as closed, until or unless it becomes raised as a problem for others.
